### PR TITLE
fix(ext/node): enable reuseport for net Socket on macOS

### LINKED ISSUE #18301: node: Error when using `npm:portfinder`
I'm having trouble running [portfinder](https://www.npmjs.com/package/portfinder), a package that finds available ports, on Deno.

Running the code below in Node.js correctly finds available ports.

```ts
import { getPortPromise } from "npm:portfinder";
// import { getPortPromise } from "portfinder";

console.log(await getPortPromise());
```

However, when I run it on Deno, I get the following error:

```shell
> deno run -A ./tmp.mjs
error: Uncaught Error: No open ports available
    at Server.onError (file:///C:/Users/ayame/AppData/Local/deno/npm/registry.npmjs.org/portfinder/1.0.32/lib/portfinder.js:58:23)
    at Object.onceWrapper (https://deno.land/std@0.168.0/node/_events.mjs:504:26)
    at Server.emit (https://deno.land/std@0.168.0/node/_events.mjs:379:28)
    at _emitErrorNT (https://deno.land/std@0.168.0/node/net.ts:1806:10)
    at processTicksAndRejections (https://deno.land/std@0.168.0/node/_next_tick.ts:41:15)
```

This problem should affect [many packages that depend on portfinder](https://www.npmjs.com/browse/depended/portfinder).

```
> deno --version
deno 1.29.0 (release, x86_64-pc-windows-msvc)
v8 10.9.194.5
typescript 4.9.4
```

### LINKED ISSUE #20618: Support VuePress (dev server)
**Platform**

Darwin 22.6.0 arm64 arm

**Version**

deno 1.40.1

**Repro:**

`deno run -A npm:create-vuepress-site`
`deno task dev`

**Expected**

Run dev server

**Actual**

```rust
➜ deno task dev
Warning Currently only basic package.json `scripts` are supported. Programs like `rimraf` or `cross-env` will not work correctly. This will be fixed in an upcoming release.
Task dev vuepress dev src
wait Extracting site metadata...
tip Apply theme @vuepress/theme-default ...
warning An error was encountered in plugin "@vuepress/plugin-back-to-top"
warning An error was encountered in plugin "@vuepress/plugin-medium-zoom"
tip Apply plugin container (i.e. "vuepress-plugin-container") ...
tip Apply plugin @vuepress/register-components (i.e. "@vuepress/plugin-register-components") ...
tip Apply plugin @vuepress/active-header-links (i.e. "@vuepress/plugin-active-header-links") ...
tip Apply plugin @vuepress/search (i.e. "@vuepress/plugin-search") ...
tip Apply plugin @vuepress/nprogress (i.e. "@vuepress/plugin-nprogress") ...
Error: listen UNKNOWN: unknown error 192.168.0.115:52759
    at __node_internal_captureLargerStackTrace (ext:deno_node/internal/errors.ts:91:9)
    at __node_internal_uvExceptionWithHostPort (ext:deno_node/internal/errors.ts:122:10)
    at Server._setupListenHandle [as _listen2] (node:net:1179:16)
    at _listenInCluster (node:net:1011:12)
    at doListen (node:net:1021:7)
    at processTicksAndRejections (ext:deno_node/_next_tick.ts:36:15)
    at runNextTicks (ext:deno_node/_next_tick.ts:71:3)
    at eventLoopTick (ext:core/01_core.js:189:21)
```

### PR DESCRIPTION
**Pull request by littledivy**
- Enables SO_REUSEPORT on macOS
- All Node `net.Socket` TCP sockets are set to `SO_REUSEPORT`.

Fixes #20618 
Fixes #18301 

Modules working: `npm:http-server`, `npm:portfinder` and Vuepress dev server

### CODE REVIEW

File: ext/net/01_net.js [hunk 1]
```diff
@@ -419,13 +419,17 @@ class Datagram {
 
 const listenOptionApiName = Symbol("listenOptionApiName");
 
-function listen(args) {
+function listen(args, checkUnstable = true) {
```
Comment (lucacasonato) [inline ext/net/01_net.js:422]: This is a public API change?

Comment (bartlomieju) [reply ext/net/01_net.js:422]: Yeah, we can't do it; we need to add another `op_net_listen_tcp` op that skips the unstable check. And we should stabilize the `reusePort` option in the next minor.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
https://github.com/denoland/deno/pull/20869/files#r1352540623

**Review by lucacasonato — CHANGES_REQUESTED** [review]
Also I think you don't want SO_REUSEPORT, but SO_REUSEADDR (and enable that by default everywhere). libuv does not enable (or let you enable) SO_REUSEPORT on TCP sockets anywhere (including macOS). It does enable SO_REUSEADDR on all Unix.

File: ext/net/01_net.js (hunk 1 above)
Comment (lucacasonato) [reply ext/net/01_net.js:422]: This is `Deno.listen`

**Comment by littledivy** [comment]
@lucacasonato  is right. libuv does not set SO_REUSEPORT for TCP (UDP only on BSDs)

It looks like portfinder is also flaky/broken in Node -

- https://github.com/http-party/node-portfinder/pull/149
- https://github.com/http-party/node-portfinder/pull/105 (Fix was rejected because they don't want to modify the library)

@bartlomieju How should we proceed?

**Comment by bartlomieju** [comment]
@littledivy I'm not sure yet - but given this package is popular we need to come up with some solution that will work in most cases (compared to always crashing right now).

**Comment by littledivy** [comment]
Closing as stale.

**Pull request closed without merging**